tracing-opentelemetry/src/layer.rs

```diff
@@ -1386,4 +1477,184 @@ mod tests {
+            .with(layer().with_tracer(tracer.clone()).max_events_per_span(4));
+        let subscriber = Arc::new(subscriber);
+
+        tracing::subscriber::with_default(subscriber.clone(), || {
+            let span = tracing::debug_span!("request");
+            let _span_guard = span.enter();
+
+            // we should record all 4 events
+            tracing::info!("event 1");
+            tracing::info!("event 2");
+            tracing::info!("event 3");
+            tracing::info!("event 4");
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // we shouldn't have recoreded any dropped events
+            assert!(extensions.get::<DroppedOtelEvents>().is_none());
+
+            // we should have all 4 events
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            assert_eq!(otel_events.len(), 4);
+
+            drop(extensions);
+            drop(data);
+
+            // record a 5th event, which is over our max of 4
+            tracing::info!("event 5");
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // since we're over our limit we should start dropping events
+            let dropped_events = extensions.get::<DroppedOtelEvents>().unwrap();
+            assert_eq!(dropped_events.dropped_count, 3);
+
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            // dropped 3, pushed one more
+            assert_eq!(otel_events.len(), 2);
+
+            drop(extensions);
+            drop(data);
+
+            // record many events!
+            for i in 6..99 {
+                tracing::info!("event {}", i);
+            }
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // we should have dropped a ton of events, and recorded that we did
+            let dropped_events = extensions.get::<DroppedOtelEvents>().unwrap();
+            assert_eq!(dropped_events.dropped_count, 96);
+
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            assert_eq!(otel_events.len(), 2);
+
+            // even though we pushed a ton of events, the capacity of our buffer
+            // should still be small
+            assert_eq!(otel_events.capacity(), 8);
+        });
+
+        // on close we should report the number of events we dropped
+        let attributes = tracer.with_data(|data| data.builder.attributes.as_ref().unwrap().clone());
+        let dropped_count_value = attributes.get(&Key::new("dropped_event_count")).unwrap();
```

Comment:
can you also test that this attribute doesn't show up if there are no dropped events?

Reply:
added!